keps/sig-storage/20200709-pod-pvc-qos-limits.md

```diff
@@ -0,0 +1,255 @@
+
+## Summary
+
+Better resource management for iops or bps of block device in pvc.
+In some senarios, we need to limit PV's iops limit or Pod iops on a device or volume at runtime.
+
+For container runtime, dockerd provides blkio related params in docker run to limit a device's iops and bps. [docker reference #block-io-bandwidth-blkio-constraint](https://docs.docker.com/engine/reference/run/#block-io-bandwidth-blkio-constraint)
+
+
+## Motivation
+The proposal would benifit in below two senarios.
+
+* Local Storage(local disk devices), better speed:
+For instances, I want to use local storage(local device) to gain the storage local speed.
+These PVCs in my cluster are blocking devices.
+
+* rootfs (preventing being overused by some pods)
+Besides, the rootfs is shared by all pods in your node, the limitation will help you make node more stable. When one pod in your node heavily uses rootfs, it may effect other pods on the same node. The limitation would help as well.
+
+
+### Goals
+
+- The limit is runtime limitation for block device when it is mounted to the pod. 
+- Only blocking device will be limited with specified volume device id. 
+- This should be implemented with cgroup, so it will only work beyond cgroup capability.
+- An extended feature that can be provided later is that we can limit iops of rootfs for container runtime.  This will reduce the influcences between pods， running on the same host.
+
+
+### Non-Goals
+
+- The limit is runtime limitation for block device when it is mounted to the pod. The limit is not a volume limitation in IaaS's aspect. If the device is used by multi pods, each pod should limit the iops by itself. 
+- For volume capability of iops, it is a physical limitation on device(PV), and this is not the same as the QoS of PVC in this proposal.
+- As cgroup implement has its limitations, we will not mention kernel buffered writings issues with cgroup limitations. This should be fixed or optimized in kernel side. Detailes will be mentioned in the `risks and limitations` below.
+
+
+## Proposal
+
+After open a feature gate "VolumeQos=True" or "DeviceIOPS=true", pod volume qos can be added by annotation like below. The alpha MVP would use annotations.
+
+
+```
+annotations:
+    qos.volume.storage.daocloud.io: >-
+        {"pvc": "snap-03", "iops": {"read": 2000, "write": 1000}, "bps":
+        {"read": 1000000, "write": 1000000}}
+```
+
+
+Then kubelet can get the mount point of the pvc and the device id. Then we use the cgroup to limit iops and bps of the pod.
+We can just edit the cgroup limit files under the pod /sys/fs/cgroup/blkio/kubepods/pod/<Container_ID>/...
+For instance, to limit read iops of a pod
+
+
+```
+ echo "<block_device_maj:min> <value>" > /sys/fs/cgroup/blkio/kubepods/pod<UID>/blkio.throttle.read_iops_device
+```
+
+To manage the QoS of containers, it can be supported in the future if there are more than 1 contaienrs in the pod, and we may add different iops limits for each container.
```

Comment:
We have discussed having both container-level and pod-level control for other resources in the past, but it can be quite complex: https://github.com/kubernetes/enhancements/pull/1592.  I would recommend assuming we will only get _either_ pod or container-level limiting.